hummingbot/connector/derivative/phemex_perpetual/phemex_perpetual_api_order_book_data_source.py

```diff
@@ -0,0 +1,234 @@
+                "update_id": snapshot_response["sequence"],
+                "bids": snapshot_response[self._diff_messages_queue_key]["bids"],
+                "asks": snapshot_response[self._diff_messages_queue_key]["asks"],
+            },
+            timestamp=snapshot_response["timestamp"] * 1e-9,
+        )
+        return snapshot_msg
+
+    async def _connected_websocket_assistant(self) -> WSAssistant:
+        url = web_utils.wss_url(CONSTANTS.PUBLIC_WS_ENDPOINT, self._domain)
+        ws: WSAssistant = await self._api_factory.get_ws_assistant()
+        async with self._api_factory.throttler.execute_task(limit_id=CONSTANTS.WSS_CONNECTION_LIMIT_ID):
+            await ws.connect(ws_url=url, ping_timeout=CONSTANTS.WS_HEARTBEAT)
+        return ws
+
+    async def _subscribe_channels(self, ws: WSAssistant):
+        """
+        Subscribes to the trade events and diff orders events through the provided websocket connection.
+        :param ws: the websocket assistant used to connect to the exchange
+        """
+        try:
+            stream_id_channel_pairs = [
+                CONSTANTS.DIFF_STREAM_METHOD,
+                CONSTANTS.TRADE_STREAM_METHOD,
+                CONSTANTS.FUNDING_INFO_STREAM_METHOD,
+            ]
+            for stream_method in stream_id_channel_pairs:
+                params = []
+                for trading_pair in self._trading_pairs:
+                    params.append(await self._connector.exchange_symbol_associated_to_pair(trading_pair=trading_pair))
+                payload = {
+                    "id": 0,
+                    "method": stream_method,
+                    "params": params if stream_method is not CONSTANTS.FUNDING_INFO_STREAM_METHOD else [],
+                }
+                subscribe_request: WSJSONRequest = WSJSONRequest(payload)
+                async with self._api_factory.throttler.execute_task(limit_id=CONSTANTS.WSS_MESSAGE_LIMIT_ID):
+                    await ws.send(subscribe_request)
+            self.logger().info("Subscribed to public order book, trade and funding info channels...")
+            safe_ensure_future(self.ping_loop(ws=ws))
+        except asyncio.CancelledError:
+            raise
+        except Exception:
+            self.logger().exception("Unexpected error occurred subscribing to order book trading and delta streams...")
+            raise
+
+    def _channel_originating_message(self, event_message: Dict[str, Any]) -> str:
+        channel = ""
+        if self._diff_messages_queue_key in event_message:
+            channel = self._diff_messages_queue_key
+        elif self._trade_messages_queue_key in event_message:
+            channel = self._trade_messages_queue_key
+        elif event_message.get("method", None) == self._funding_info_messages_queue_key:
+            channel = self._funding_info_messages_queue_key
+        elif event_message.get("result", None) == "pong":
+            self.pong_received_event.set()
```

Comment:
Processing the ping message should be done in the `_process_message_for_unknown_channel` method. The method `_channel_originating_message` should only have logic to determine the channel a message belongs to

Reply:
done